server/storage/wal/wal_test.go

```diff
@@ -1180,3 +1180,203 @@ func TestLastRecordLengthExceedFileEnd(t *testing.T) {
+				wantEntries: []raftpb.Entry{{Index: 1, Data: []byte("a")}, {Index: 2, Data: []byte("b")}, {Index: 3, Data: []byte("b")}, {Index: 4, Data: []byte("c")}, {Index: 5, Data: []byte("c")}},
+			},
+			readAllEntries: want{
+				wantState:   raftpb.HardState{Commit: 4},
+				wantEntries: []raftpb.Entry{{Index: 1, Data: []byte("a")}, {Index: 2, Data: []byte("b")}, {Index: 3, Data: []byte("b")}, {Index: 4, Data: []byte("c")}, {Index: 5, Data: []byte("c")}},
+			},
+		},
+		{
+			name: "before snapshot",
+			operations: []batch{
+				{
+					state:   &raftpb.HardState{Commit: 1},
+					entries: []raftpb.Entry{{Index: 1, Data: []byte("a")}, {Index: 2, Data: []byte("b")}},
+				},
+				{
+					snapshot: &walpb.Snapshot{Index: 3, ConfState: &raftpb.ConfState{}},
+				},
+				{
+					state:   &raftpb.HardState{Commit: 5},
+					entries: []raftpb.Entry{{Index: 4, Data: []byte("d")}, {Index: 5, Data: []byte("e")}},
+				},
+			},
+			walReadAll: want{
+				wantError: "slice bounds out of range",
+			},
+			readAllEntries: want{
+				wantState:   raftpb.HardState{Commit: 5},
+				wantEntries: []raftpb.Entry{{Index: 1, Data: []byte("a")}, {Index: 2, Data: []byte("b")}, {Index: 4, Data: []byte("d")}, {Index: 5, Data: []byte("e")}},
+			},
+		},
+		{
+			name: "at snapshot",
+			operations: []batch{
+				{
+					state:   &raftpb.HardState{Commit: 1},
+					entries: []raftpb.Entry{{Index: 1, Data: []byte("a")}, {Index: 2, Data: []byte("b")}},
+				},
+				{
+					snapshot: &walpb.Snapshot{Index: 3, ConfState: &raftpb.ConfState{}},
+				},
+				{
+					state:   &raftpb.HardState{Commit: 5},
+					entries: []raftpb.Entry{{Index: 4, Data: []byte("d")}, {Index: 5, Data: []byte("e")}},
+				},
+			},
+			readAt: walpb.Snapshot{Index: 3},
+			walReadAll: want{
+				wantState:   raftpb.HardState{Commit: 5},
+				wantEntries: []raftpb.Entry{{Index: 4, Data: []byte("d")}, {Index: 5, Data: []byte("e")}},
+			},
+			readAllEntries: want{
+				wantState:   raftpb.HardState{Commit: 5},
+				wantEntries: []raftpb.Entry{{Index: 1, Data: []byte("a")}, {Index: 2, Data: []byte("b")}, {Index: 4, Data: []byte("d")}, {Index: 5, Data: []byte("e")}},
+			},
+		},
+		{
+			name: "after snapshot",
+			operations: []batch{
+				{
+					state:   &raftpb.HardState{Commit: 1},
+					entries: []raftpb.Entry{{Index: 1, Data: []byte("a")}, {Index: 2, Data: []byte("b")}},
+				},
+				{
+					snapshot: &walpb.Snapshot{Index: 3, ConfState: &raftpb.ConfState{}},
+				},
+				{
+					state:   &raftpb.HardState{Commit: 5},
+					entries: []raftpb.Entry{{Index: 4, Data: []byte("d")}, {Index: 5, Data: []byte("e")}},
+				},
+			},
+			readAt: walpb.Snapshot{Index: 4},
+			walReadAll: want{
+				wantError: "snapshot not found",
+			},
+			readAllEntries: want{
+				wantState:   raftpb.HardState{Commit: 5},
+				wantEntries: []raftpb.Entry{{Index: 1, Data: []byte("a")}, {Index: 2, Data: []byte("b")}, {Index: 4, Data: []byte("d")}, {Index: 5, Data: []byte("e")}},
+			},
+		},
+	}
```

Comment:
some more test cases to consider:
 - to test the logic on line 424, make the first entry index something like `5` and then subsequent entries with their index less than or equal to it.  we should expect that those subsequent entries get ignored by `ReadAllEntries`.
 - not sure how complicated this would be since `ReadAll` doesn't have the same logic like line 449, but we should test that the function exits immediately when it fails to decode mid way through the entries.

Reply:
Added test for the first case, for the second this test assumes correct WAL. Handling corrupted WAL is out of scope. That's testing of CRC mechanism.